shell/platform/linux/fl_texture_registrar.cc

```diff
@@ -0,0 +1,136 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_texture_registrar.h"
+
+#include <gmodule.h>
+
+#include "flutter/shell/platform/embedder/embedder.h"
+#include "flutter/shell/platform/linux/fl_engine_private.h"
+#include "flutter/shell/platform/linux/fl_texture_private.h"
+#include "flutter/shell/platform/linux/fl_texture_registrar_private.h"
+
+struct _FlTextureRegistrar {
+  GObject parent_instance;
+
+  FlEngine* engine;
+
+  GHashTable* textures;
+};
+
+G_DEFINE_TYPE(FlTextureRegistrar, fl_texture_registrar, G_TYPE_OBJECT)
+
+static void engine_weak_notify_cb(gpointer user_data,
+                                  GObject* where_the_object_was) {
+  FlTextureRegistrar* self = FL_TEXTURE_REGISTRAR(user_data);
+  self->engine = nullptr;
+
+  // Unregister any textures.
+  g_autoptr(GHashTable) textures = self->textures;
+  self->textures = g_hash_table_new_full(g_direct_hash, g_direct_equal, nullptr,
+                                         g_object_unref);
+  g_hash_table_remove_all(textures);
+}
+
+static void fl_texture_registrar_dispose(GObject* object) {
+  FlTextureRegistrar* self = FL_TEXTURE_REGISTRAR(object);
+
+  g_clear_pointer(&self->textures, g_hash_table_unref);
+
+  if (self->engine != nullptr) {
+    g_object_weak_unref(G_OBJECT(self->engine), engine_weak_notify_cb, self);
+    self->engine = nullptr;
+  }
+
+  G_OBJECT_CLASS(fl_texture_registrar_parent_class)->dispose(object);
+}
+
+static void fl_texture_registrar_class_init(FlTextureRegistrarClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = fl_texture_registrar_dispose;
+}
+
+static void fl_texture_registrar_init(FlTextureRegistrar* self) {
+  self->textures = g_hash_table_new_full(g_direct_hash, g_direct_equal, nullptr,
+                                         g_object_unref);
+}
+
+G_MODULE_EXPORT int64_t
+fl_texture_registrar_register_texture(FlTextureRegistrar* self,
+                                      FlTexture* texture) {
+  g_return_val_if_fail(FL_IS_TEXTURE_REGISTRAR(self), 0);
+  g_return_val_if_fail(FL_IS_TEXTURE(texture), 0);
+
+  int64_t id = fl_texture_get_texture_id(texture);
+  g_hash_table_insert(self->textures, GINT_TO_POINTER(id),
+                      g_object_ref(texture));
+
+  if (self->engine == nullptr) {
```

Comment:
This is confusing, because we've actually assigned an ID and taken a reference to the texture, but then we return 0. This should probably be:

```c
int64_t id = fl_texture_get_texture_id(texture);
g_hash_table_insert(self->textures, GINT_TO_POINTER(id), g_object_ref(texture));

if (self->engine != nullptr) {
 fl_engine_register_external_texture(self->engine, id);
}

return id;
```

Reply:
Or the following is more consistent with the other code in this file:
```c
if (self->engine == nullptr) {
  return id;
}
```